nixos/lib/make-disk-image.nix

```diff
@@ -235,12 +255,45 @@ let format' = format; in let
 
     ${if diskSize == "auto" then ''
       ${if partitionTableType == "efi" || partitionTableType == "hybrid" then ''
-        additionalSpace=$(( ($(numfmt --from=iec '${additionalSpace}') + $(numfmt --from=iec '${bootSize}')) / 1000 ))
+        # Add the GPT at the end
+        gptSpace=$(( 512 * 34 * 1 ))
+        # Normally we'd need to account for alignment and things, if bootSize
+        # represented the actual size of the boot partition. But it instead
+        # represents the offset at which it ends.
+        # So we know bootSize is the reserved space in front of the partition.
+        reservedSpace=$(( gptSpace + $(numfmt --from=iec '${bootSize}') ))
+      '' else if partitionTableType == "legacy+gpt" then ''
+        # Add the GPT at the end
+        gptSpace=$(( 512 * 34 * 1 ))
+        # And include the bios_grub partition; the ext4 partition starts at 2MB exactly.
+        reservedSpace=$(( gptSpace + 2 * 1024*1024 ))
+      '' else if partitionTableType == "legacy" then ''
+        # Add the 1MiB aligned reserved space (includes MBR)
+        reservedSpace=$(( 1024*1024 ))
       '' else ''
-        additionalSpace=$(( $(numfmt --from=iec '${additionalSpace}') / 1000 ))
+        reservedSpace=0
       ''}
-      diskSize=$(( $(set -- $(du -d0 $root); echo "$1") + $additionalSpace ))
-      truncate -s "$diskSize"K $diskImage
+      additionalSpace=$(( $(numfmt --from=iec '${additionalSpace}') + reservedSpace ))
+
+      # Compute required space in filesystem blocks
+      diskUsage=$(find . ! -type d -exec 'du' '--apparent-size' '--block-size' "${blockSize}" '{}' ';' | cut -f1 | sum_lines)
+      # Each inode takes space!
+      numInodes=$(find . | wc -l)
+      # Convert to bytes, inodes take two blocks each!
+      diskUsage=$(( (diskUsage + 2 * numInodes) * ${blockSize} ))
+      # Then increase the required space to account for the reserved blocks.
+      fudge=$(compute_fudge $diskUsage)
+      requiredFilesystemSpace=$(( diskUsage + fudge ))
+
+      diskSize=$(( requiredFilesystemSpace  + additionalSpace ))
```

Comment:
This isn't necessarily a multiple of 512 bytes and makes fdisk/gdisk complain. parted seems to ignore it, but probably still a good idea to round up to the next highest blockSize here.

Reply:
This probably would be wrong.

What I assume it complains about is the partitions starting/ending unaligned, and not necessarily on sector size (commonly 512 bytes). I believe the alignment by default is 1MiB? Though alignment is actually a property of the block device being partitioned! Doing any alignment on an image might actually end up wrong!

For this, we'd need to re-do the whole disk image build logic to actually work with discrete filesystem images that we *then* place as partitions into a larger disk image.

Right now, we're actually trying to carve a hole into a bigger image we *hope* we're making big enough. Well, we're doing less of it with this PR.

Though I guess rounding up to the nearest megabyte is a clean way to avoid this warning.

Reply:
It's actually complaining about file size (three whole times for some reason). And it's just gdisk.

```
$ truncate -s 1234 test
$ gdisk test
GPT fdisk (gdisk) version 1.0.5

Warning: File size is not a multiple of 512 bytes! Misbehavior is likely!
Warning: File size is not a multiple of 512 bytes! Misbehavior is likely!
Warning: File size is not a multiple of 512 bytes! Misbehavior is likely!
```

Yeah, rounding up to the nearest MB would fix it. I think parted actually recommends that everything be 4MiB aligned at least. Best we can do is guess here anyway and 1 MB alignment seems good enough for most cases.

Reply:
Right, this warning is probably a non-issue if it is on a disk image:

 - https://sourceforge.net/p/gptfdisk/code/ci/eae9f7e2e1470ed2d5bc029180aa2adc88fbc4a7/tree/diskio-unix.cc#l496
 - https://sourceforge.net/p/gptfdisk/code/ci/eae9f7e2e1470ed2d5bc029180aa2adc88fbc4a7/tree/diskio-unix.cc#l493

Hmm... Though it would mean that the backup GPT at the end of the disk is not at LBAEND-33, but rather LBAEND-34, since it rounded down the amount of sectors. BUT, again this is a non-issue since when `dd`'ing the disk image to the actual device, the device will always be bigger, and that the backup GPT will invariably not be at the end sector.

Reply:
And it is done

https://github.com/NixOS/nixpkgs/pull/120489/commits/7b8b3fab6dce5357c14fc435b436514fc985f0f7